### ISSUE TITLE
Dagger-producers 2.0-beta has compile dependency on google guava 18

### ISSUE THREAD

**Issue by kivsiak** [google/dagger#233]
Here is a part of grade dependency query:
 
```
compile - Classpath for compiling the main sources.
+--- com.android.support:appcompat-v7:23.0.1
|    \--- com.android.support:support-v4:23.0.1
|         \--- com.android.support:support-annotations:23.0.1
+--- com.google.dagger:dagger:2.0.1
|    \--- javax.inject:javax.inject:1
+--- com.google.dagger:dagger-producers:2.0-beta
|    +--- com.google.dagger:dagger:2.0 -> 2.0.1 (*)
|    \--- com.google.guava:guava:18.0
```

Guava with 14k methods is very undesirable dependency for android. Will it be removed in future releases?

**Comment by JakeWharton** [comment]
`ListenableFuture` is the foundation of producers. If you must use producers, use proguard.

**Comment by divath** [comment]
Please make it happen, I don't understand why we have to expose it.

**Comment by JakeWharton** [comment]
Quite obviously because the feature is built on Guava's ListenableFutures. Without it, there is no asynchronous execution mechanism that composes.

**Comment by divath** [comment]
But cannot guava dependency be shaded like it is in autovalue?
All stuff matching com.google is moved to autovalue.shaded.com.google.common for example.
(https://github.com/google/auto/blob/master/value/pom.xml)

**Comment by JakeWharton** [comment]
No. Guava types are used in the public API and that would prevent interoperability with normal usage of ListenableFutures by other libraries and code.

**Comment by cjp39** [comment]
ListenableFuture has been around since 1.0. There's no need to force external users to use the latest version of Guava (losing backwards compatibility since Guava removes methods periodically), simply because you happen to use version 18 features internally. Shading would decouple that.

Additionally, given the totally legitimate concern of kivsiak, I would argue that you should be moving the Producers API into an optional JAR, so Guava is not forced onto people who just want Dagger.

**Comment by JakeWharton** [comment]
It already is a separate jar.

**Comment by cjp39** [comment]
So you can shade Dagger then.

**Comment by JakeWharton** [comment]
No. Guava types are used in the public API and that would prevent interoperability with normal usage of ListenableFutures by other libraries and code.

**Comment by cjp39** [comment]
Are any of those Guava types only available in v17?

**Comment by cjp39** [comment]
Otherwise, you're conflating the Guava Dagger uses internally with the Guava Dagger asks to be linked against. They're not the same, even though internally Dagger happens to assume they are when it does things like pass ListenableFuture.class to code writers.

**Comment by cjp39** [comment]
And you can (and should) shade the Guava uses at compile time.

**Comment by JakeWharton** [comment]
Oooh, yeah sure. The compiler could use shaded Guava and use only TypeMirrors / ClassName to reference ListenableFuture in the parsed and generated code.

**Comment by cjp39** [comment]
If you want a quicker short-term fix, you could strip the shade prefix from classes before you write them. Up to you whether you think that's a sensible long-term solution. Passing in Foo.class sure is convenient.

**Comment by gk5885** [comment]
Sorry this escaped notice for a bit.  A few thoughts here.

First, @JakeWharton is absolutely correct.  `ListenableFuture` is the foundation of producers, and particularly the listenable bit -- it is critical for the efficient scheduling of the tasks in the graph.  Maybe someday we can standardize on [`CompletableFuture`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html), but for the time being, Guava is required.

Second, Dagger 2 on Android more-or-less assumes [ProGuard](http://proguard.sourceforge.net/).  That is actually about to be called out in the documentation explicitly.  There are a variety of benefits that might not even be particularly obvious, but getting rid of most of Guava seems like a pretty decent motivator.

But of course, we can examine the dependency itself.  The interesting bit is how `ListenableFuture` relates to the user-facing API (@jbeder to check my assessment here).  If it were the case that it were only used in the implementation, then we could shade things without issue, but there are a handful of critical places where users need it:

1. By injecting a `ListenableFuture` you are providing a signal to the framework that the task should have been started, but that the given `@Produces` method should be allowed to begin executing _before_ it has necessarily completed (unlike injecting an instance or `Produced`).  It's not particularly common, but is a pattern that occasionally arises.
2. By injecting a `Provider`.  This is essentially conditionally exercising a binding.  E.g. a produces method needs to collect debug data, but only if a request has been flagged as a debug request.  Of course, the way to model a task that is currently executing is a `ListenableFuture`.  Now, this is a situation where we _could potentially_ get away with just returning `Future` since the most common operation would be to just call `get()`, but it is extraordinarily difficult and wasteful to transform a `Future` into a `ListenableFuture`, so it would be very limiting if you happened to be using other libraries that needed `ListenableFuture` as well.
3. By returning a `ListenableFuture` from a `@Produces` method.  This is essential to how you integrate non-producer code (e.g. rpc frameworks) into a producer graph.  Without this mechanism, you have to burn a thread in the producer's executor waiting on an operation that is already asynchronous.  :(

So, the question is whether or not those 3 use cases are essential to the framework.  If they're not, you could argue that another "flavor" of dagger producers that didn't have those features could exist for just this case, but at the moment they feel pretty central to me.

Finally, I assigned @cgruber to this bug since I know there have been grumblings about how we package Guava w.r.t. Android and he might have some more build-level insight.

**Comment by cjp39** [comment]
@gk5885 As mentioned above, the Guava library used internally by the Dagger compiler can be shaded without affecting the external behaviour. Indeed, the Dagger compiler could support ListenableFuture **without using Guava at all internally**. All it needs is to find references to com.google.common.util.concurrent.ListenableFuture, which it can do by hard-coding that path. *As it happens*, because those classes happen to be around anyway, the Dagger compiler is calling ListenableFuture.getName() to avoid hard-coding those qualified names as strings. This breaks when Dagger's internal Guava library gets shaded. However, there's no need to do it this way, and other code builders do not: for instance, FreeBuilder has support for Jackson annotations, but isn't even compiled against it, let alone shipped with it. Equally, if you wanted to support CompletableFuture, there would be no need to compile Dagger against Java 8. It can just find references to java.util.concurrent.CompletableFuture, while remaining completely backwards-compatible.

So whether Dagger shades Guava in its compiler or not has nothing to do with whether the Dagger API uses Guava types or not, in Producers or anywhere else.

**Comment by tbroyer** [comment]
@cjp39 Re:
> There's no need to force external users to use a really recent version of Guava (losing backwards compatibility since Guava removes methods periodically), simply because you happen to use version 17 features internally.

Dagger only forces Guava 17+ in the `processorpath`, not in the project itself.

Re. "Guava removes methods periodically", methods have to be [deprecated for 2 years before they're removed](https://github.com/google/guava/blob/master/README.md#important-warnings) (note that this is part of an “important warnings” section), so that'd mean you use a "really old" version of Guava (or you used `@Beta` methods). IMO, if you use Guava, you should play by its rules, and update to a newer Guava version at least every 2 years (more frequently if you use `@Beta` features); if you're not going to do that, you should (IMO) find another library.
In other words, if Guava 17+ is a problem for you, then You're Doing It Wrong™ (IMO), and if the breakage comes from a library, then the library author is doing it wrong ⇒ file a bug on the library so they update Guava.

…and BTW the original issue was about “Guava with 14k methods is very undesirable dependency for android”, i.e. about runtime dependency.

**Comment by JakeWharton** [comment]
Yes, I have become supportive of this change since my original comment, as
I've ran into this problem in other libraries. Referencing classes in the
processor and assuming they will have equivalent names and packages isn't a
sound strategy.

**Comment by cjp39** [comment]
@tbroyer 1. The processorpath overrides the compile path if the JARs have the same name.
2. Breaking old code is simply unnecessary in this case, irrespective of what the Guava site says.
3. You want to create multiple GitHub issues for the same root cause? Will that help anyone? Surely we'd just have the same discussion in multiple places?

**Comment by JakeWharton** [comment]
Oops I replied to wrong issue!

**Comment by cjp39** [comment]
Have I horribly mixed up two issues?

**Comment by cjp39** [comment]
Do we want a new issue for the shading?

**Comment by gk5885** [comment]
@cjp39 This issue (as originally stated) is not about the use of Guava in the processor.  My response was meant to address the issue that was filed, which is the strong dep from the producers API to Guava.  

You seem to have a separate concern about old versions of Guava.  Versions earlier than 17 predate Dagger 2 itself (Guava 17 was published to Maven on 22-Apr-2014).  Supporting them is a non-goal, given that, as @tbroyer mentioned, we have a very carefully cultivated policy for ensuring that swapping in a newer version will not break code compiled against older versions as long as you're within a very large window.

If you have other concerns relating to the Guava dep in the processor, we'd be open to discussing them, but please file a separate issue.

@JakeWharton We don't _assume_ that the Guava classes will have equivalent names; it is explicitly guaranteed by Guava in the doc referenced by @tbroyer.  Were it any other library without those compatibility guarantees, I would agree, but since there is a strict policy in place, the indirection adds unnecessary complexity.
